### PR TITLE
Zoom pan behavior

### LINKED ISSUE #15399: Zoom tool, assign right-click
Author Name: **andskog -** (andskog -)
Original Redmine Issue: [5984](https://issues.qgis.org/issues/5984)

Redmine category:map_canvas

---
When zooming with the magnifying glass (+ and -), the left mouse button is assigned with the obvious function, but why not assign the right button too, to do the opposite zoom function?

### LINKED ISSUE #52461: Zooming is not centered around mouse but is offset
### What is the bug or the crash?

When zooming in our out, map position is offset (see video below)

### Steps to reproduce the issue

https://user-images.githubusercontent.com/22519832/229078247-d27e385e-adc7-4e42-b786-049bcd43c85f.mov

### Versions

Using MacOS

QGIS-Version | 3.28.2-Firenze | QGIS-Codeversion | b47e00ba60
-- | -- | -- | --
Qt-Version | 5.15.2
Python-Version | 3.9.5
GDAL-Version | 3.3.2
PROJ-Version | 8.1.1
EPSG-Registraturdatenbankversion | v10.028 (2021-07-07)
GEOS-Version | 3.9.1-CAPI-1.14.2
SQLite-Version | 3.35.2
PDAL-Version | 2.3.0
PostgreSQL-Client-Version | unknown
SpatiaLite-Version | 5.0.1
QWT-Version | 6.1.6
QScintilla2-Version | 2.11.5
BS-Version | macOS 13.0
  |   |   |  
Aktive Python-Erweiterungen
QuickOSM | 2.1.1
wfsclient | 0.9.11
processing | 2.12.99
sagaprovider | 2.12.99
grassprovider | 2.12.99
db_manager | 0.1.20
MetaSearch | 0.3.6

</body></html>QGIS-Version
3.28.2-Firenze
QGIS-Codeversion
[b47e00ba60](https://github.com/qgis/QGIS/commit/b47e00ba60)
Qt-Version
5.15.2
Python-Version
3.9.5
GDAL-Version
3.3.2
PROJ-Version
8.1.1
EPSG-Registraturdatenbankversion
v10.028 (2021-07-07)
GEOS-Version
3.9.1-CAPI-1.14.2
SQLite-Version
3.35.2
PDAL-Version
2.3.0
PostgreSQL-Client-Version
unknown
SpatiaLite-Version
5.0.1
QWT-Version
6.1.6
QScintilla2-Version
2.11.5
BS-Version
macOS 13.0

Aktive Python-Erweiterungen
QuickOSM
2.1.1
wfsclient
0.9.11
processing
2.12.99
sagaprovider
2.12.99
grassprovider
2.12.99
db_manager
0.1.20
MetaSearch
0.3.6

### Supported QGIS version

- [X] I'm running a supported QGIS version according to the roadmap.

### New profile

- [X] I tried with a new QGIS profile

### Additional context

_No response_

### PR DESCRIPTION
**Pull request by YoannQDQ**
- Fixes #52461

## Description

This PR aims to make the zoom and pan tools more intuitive.

### ZoomIn/Out tools improvements
- Zoom In/Out keep the clicked point under mouse by default (like the wheel zoom). To zoom and center on the clicked point (previous default behavior), press Ctrl

### PanTool improvements
- When single clicking, pan to the clicked point only if Ctrl is pressed.
- Double clicking the canvas while the pan tool is active zooms in while keeping the clicked point under the cursor
- From any other tool, Ctrl+Middle click pans to the clicked point.

### CODE REVIEW

**Comment by nyalldawson** [comment]
Nice work once again @YoannQDQ  !!

I have a couple of high-level thoughts before reviewing the code:

> ZoomIn/Out tools improvements
> Use the right click to invert the zoomIn/zoomOut tool (Same behavior as pressing the Alt key)

I really want to keep single right click reserved for displaying the context menu unless there's a REALLY good reason not to (eg for the digitizing tools). I'd suggest making the right click inversion only kick in when there's a drag involved...but honestly I'd love to also reserve this for a smooth camera "dolly" zoom like we use in the 3d map canvases! 

> Disable the drag zoom while zooming out, as it made little sense

Actually there IS some logic here, it's just not immediately evident! What it does is zoom out so that the current extent of the map will fit into the dragged rectangle. I don't think we should remove it -- while I don't personally use it extensively, I can easily imagine that it's a natural part of some users workflows.

**Comment by YoannQDQ** [comment]
> honestly I'd love to also reserve this for a smooth camera "dolly" zoom like we use in the 3d map canvases!

Not sure what you mean by dolly zoom in this context (for me, this means Hitchcock's Vertigo effect, but I don't see how it relates to a 2D canvas). Sure, I can keep the context menu. On a side note, the zoom and pan tool are flagged as `Transient`, which is not compatible with the `ShowContextMenu` flag, which makes the `Transient` flag effectively unsued in the code base. It might be used in plugins though, but should Transient be flagged as deprecated?

> there IS some logic here

"Makes little sense" was maybe too strong. I still think it is counter intuitive, as it makes the user draw a rect area in screen coordinates then cram the current extent into it, whereas map tools usually work on map/layers coordinates. Ok too keep it, but I'd be curious to know if this is actually part of some workflows. Wish we had some kind of telemetry data collection, but that is an entirely different (and probably controversial) subject.

**Comment by uclaros** [comment]
I +1 Nyall's comments above and would also like to add my objection to _Pan center on point only if Ctrl is pressed_. Click-to-pan is definitely on a user's workflow when needing to browse through a whole area at a fixed scale, so I wouldn't hide that behind a ctrl modifier. What's worse is that this change is done to not break the double-click-to-zoom, a functionality which is completely broken in multiple ways and I really doubt anyone uses it - I'd completely remove it.
If you really want the double-click-to-zoom, I'd put that one behind the ctrl modifier instead.

**Comment by YoannQDQ** [comment]
> What's worse is that this change is done to not break the double-click-to-zoom
 
Actually it was NOT done to keep the double click zoom ; that was just a happy side-effect. The entire point was to make the map navigation more intuitive, especially to first time users, so making the pan tool behave by default like mainstream apps (e.g. Google Maps), while keeping "advanced" feature available with modifiers seemed like a good idea. The double tap to zoom also exists in Google maps by the way, and this PR makes it "unbroken" ;)

> Click-to-pan is definitely on a user's workflow when needing to browse through a whole area at a fixed scale

What do you think about adding a setting `pan_on_single_click` (False by default) that makes it the default behavior?

**Comment by uclaros** [comment]
> @uclaros What do you think about adding a setting `pan_on_single_click` (False by default) that makes it the default behavior (no need for Ctrl modifier when active)?

I'd be OK with a setting that defaults to _**True**_, but the introduction of minor settings is generally frowned upon (and it would not apply to newcomers any more :)

One can zoom pretty intuitively with the mouse wheel, I can hardly imagine a new user struggling to find a way to zoom! Then one can zoom with the zoom in/zoom out tools, with shift+click using the pan tool, with shift+middle click using _any_ tool, by directly changing the scale, by ctr++, ... do we need more ways to zoom?
However, panning the map without moving your wrist, it's either the arrow keys (fixed offset) or the pan tool.

**Comment by YoannQDQ** [comment]
> I can hardly imagine a new user struggling to find a way to zoom!

I wholly agree with you here, but then again, the point was not to add new ways to zoom. Making pressing the CTRL modifier needed to pan to the clicked point was not an unfortunate consequence of fixing the double click feature, it was fuly intentional. I personally was a bit confused by the click to pan behavior at first, so I thought making it more explicit was a good idea. IMO having a consistent CTRL modifier behavior for the pan tool, middle click pan and zoom tool is desirable.

> I'd be OK with a setting that defaults to **_True_**, but the introduction of minor settings is generally frowned upon (and it would not apply to newcomers any more :)

Yeah, making it opt-out kinda defeats the point.

A possible way forward:

- Ditch the right click invert zoom feature. Waiting for @nyalldawson input on the Transient flag
- Restore the click and drag zoom out
- Add a new `pan_on_single_click` setting, default value True
- When `pan_on_single_click` is True, clicking the canvas while the pan tool is active recenter the map on the clicked point (same a now). When it is False, one needs to press the CTRL modifier to get the same result.   
- For the sake of consistency, make the middle click behave like the pan tool (pan on click if `pan_on_single_click` is True, pan on CTRL+click if `pan_on_single_click` is False). This effectively changes the current behavior
- If `pan_on_single_click` is True, double click zoom works only if CTRL is pressed. 
- Zoom in and zoom out do not recenter the map, except when CTRL is pressed, regardless of the `pan_on_single_click` setting.

**Comment by nyalldawson** [comment]
> Ditch the right click invert zoom feature. Waiting for @nyalldawson input on the Transient flag

I'm +1 to completely killing the transient flag. It's been broken for a while and no-one has complained. If this transient functionality is desired then we need to use something much less valuable vs the right-click button. (Eg make it a configurable setting for mice with > 3 buttons only...)

File: src/gui/qgsmaptool.h [hunk 1]
```diff
@@ -109,9 +109,6 @@ class GUI_EXPORT QgsMapTool : public QObject
      */
     enum Flag
     {
```
Comment (nyalldawson) [inline src/gui/qgsmaptool.h:111]: We will need to leave the transient value here, but change the docs to say that it's deprecated and no longer has any effect.

**Comment by nyalldawson** [comment]
@YoannQDQ can you update the summary in the pr description please?

**Comment by nyalldawson** [comment]
is this still a draft?

**Comment by YoannQDQ** [comment]
@nyalldawson

This works as it is, but this changes the cuent behavior in a way that is IMHO more intuitive, but that might go against long-established practices (see @uclaros comments).

It the current state of this PR, pan to click only occurs when the Ctrl key is pressed, whether it is from a left click with the pan tool active, or from the middle click with any other tool active.

I suggested to add an new `pan_on_single_click` setting, but I'm not really happy with this.
Any thoughts on the matter?

**Pull request closed without merging**